src/MixedQuoter.sol

```diff
@@ -206,62 +242,174 @@ contract MixedQuoter is IMixedQuoter, IPancakeV3SwapCallback {
+                    bytes memory swapListBytes = MixedQuoterRecorder.getV4PoolSwapList(poolHash);
+                    IQuoter.QuoteExactSingleParams[] memory swapHistoryList;
+                    uint256 swapHistoryListLength;
+                    if (swapListBytes.length > 0) {
+                        swapHistoryList = abi.decode(swapListBytes, (IQuoter.QuoteExactSingleParams[]));
+
+                        swapHistoryListLength = swapHistoryList.length;
+                    }
+                    IQuoter.QuoteExactSingleParams[] memory swapList =
+                        new IQuoter.QuoteExactSingleParams[](swapHistoryListLength + 1);
+                    for (uint256 i = 0; i < swapHistoryListLength; i++) {
+                        swapList[i] = swapHistoryList[i];
+                    }
+                    swapList[swapHistoryListLength] = swapParams;
+
+                    (amountIn, gasEstimateForCurAction) = binQuoter.quoteExactInputSingleList(swapList);
+                    swapListBytes = abi.encode(swapList);
+                    MixedQuoterRecorder.setV4PoolSwapList(poolHash, swapListBytes);
+                } else {
+                    (amountIn, gasEstimateForCurAction) = binQuoter.quoteExactInputSingle(swapParams);
+                }
             } else if (action == MixedQuoterActions.SS_2_EXACT_INPUT_SINGLE) {
                 (tokenIn, tokenOut) = convertNativeToWETH(tokenIn, tokenOut);
                 // params[actionIndex] is zero bytes
-                (amountIn, gasEstimateForCurAction) = quoteExactInputSingleStable(
-                    QuoteExactInputSingleStableParams({
-                        tokenIn: tokenIn,
-                        tokenOut: tokenOut,
-                        amountIn: amountIn,
-                        flag: 2
-                    })
-                );
+
+                if (!withContext) {
+                    (amountIn, gasEstimateForCurAction) = quoteExactInputSingleStable(
+                        QuoteExactInputSingleStableParams({
+                            tokenIn: tokenIn,
+                            tokenOut: tokenOut,
+                            amountIn: amountIn,
+                            flag: 2
+                        })
+                    );
+                } else {
+                    bool zeroForOne = tokenIn < tokenOut;
+                    bytes32 poolHash = MixedQuoterRecorder.getSSPoolHash(tokenIn, tokenOut);
+                    // update stable pool swap direction, only allow one direction in one transaction
+                    MixedQuoterRecorder.setAndCheckSwapDirection(poolHash, zeroForOne);
+                    (uint256 accAmountIn, uint256 accAmountOut) =
+                        MixedQuoterRecorder.getPoolSwapTokenAccumulation(poolHash, zeroForOne);
+                    uint256 swapAmountOut;
+                    amountIn += accAmountIn;
+                    (swapAmountOut, gasEstimateForCurAction) = quoteExactInputSingleStable(
```

Comment:
do we need to have "quote*WithAccumulation" version for ss pool as well ? Not familiar with it's charging rule, does the fee also auto get reinvested into the pool ?

Reply:
should not need. 
it is different from v2.